### ISSUE TITLE
LayoutLM: how to get prediction confidence scores?

### ISSUE THREAD

**Issue by sbstcano** [microsoft/unilm#270]
Hi, I'm trying to get prediction confidence scores. Anyone knows how to get them?

**Comment by sreejith3534** [comment]
def evaluate(args, model, tokenizer, labels, pad_token_label_id, mode, prefix=""):
    eval_dataset = FunsdDataset(args, tokenizer, labels, pad_token_label_id, mode=mode)

    args.eval_batch_size = args.per_gpu_eval_batch_size * max(1, args.n_gpu)
    eval_sampler = SequentialSampler(eval_dataset)
    eval_dataloader = DataLoader(
        eval_dataset,
        sampler=eval_sampler,
        batch_size=args.eval_batch_size,
        collate_fn=None,
    )

    # Eval!
    logger.info("***** Running evaluation %s *****", prefix)
    logger.info("  Num examples = %d", len(eval_dataset))
    logger.info("  Batch size = %d", args.eval_batch_size)
    eval_loss = 0.0
    nb_eval_steps = 0
    preds = None
    out_label_ids = None
    logits_store = None
    model.eval()
    for batch in tqdm(eval_dataloader, desc="Evaluating"):
        with torch.no_grad():
            inputs = {
                "input_ids": batch[0].to(args.device),
                "attention_mask": batch[1].to(args.device),
                "labels": batch[3].to(args.device),
            }
            if args.model_type in ["layoutlm"]:
                inputs["bbox"] = batch[4].to(args.device)
            inputs["token_type_ids"] = (
                batch[2].to(args.device)
                if args.model_type in ["bert", "layoutlm"]
                else None
            )  # RoBERTa don"t use segment_ids
            outputs = model(**inputs)
            tmp_eval_loss, logits = outputs[:2]

            if args.n_gpu > 1:
                tmp_eval_loss = (
                    tmp_eval_loss.mean()
                )  # mean() to average on multi-gpu parallel evaluating

            eval_loss += tmp_eval_loss.item()
        nb_eval_steps += 1
        logits_1 = F.softmax(logits, dim=2)
        logits_label = torch.argmax(logits_1, dim=2)
        logits_label = logits_label.detach().cpu().numpy().tolist()[0]
        logits_confidence = [values[label].item() for values, label in zip(logits_1[0], logits_label)]
        if preds is None:
            logits_store = np.array(logits_confidence).reshape(1, -1)
            preds = logits.detach().cpu().numpy()
            out_label_ids = inputs["labels"].detach().cpu().numpy()
        else:
            logits_store = np.append(np.array(logits_confidence).reshape(1, -1), axis=0)
            preds = np.append(preds, logits.detach().cpu().numpy(), axis=0)
            out_label_ids = np.append(
                out_label_ids, inputs["labels"].detach().cpu().numpy(), axis=0
            )
    # print('out_label_ids==',out_label_ids)
    eval_loss = eval_loss / nb_eval_steps
    preds = np.argmax(preds, axis=2)
    # # logits = F.softmax(preds, dim=2)
    # # preds = torch.argmax(preds, dim=2)
    # # logits_label = torch.from_numpy(preds).detach().cpu().numpy().tolist()[0]
    # logits_label = torch.from_numpy(np.argmax(preds,axis = 2)).detach().cpu().numpy().tolist()[0]
    # logits_confidence = [values[label].item() for values, label in zip(logits_store, logits_label)]
    label_map = {i: label for i, label in enumerate(labels)}

    out_label_list = [[] for _ in range(out_label_ids.shape[0])]
    preds_list = [[] for _ in range(out_label_ids.shape[0])]
    confidence_lst = [[] for _ in range(out_label_ids.shape[0])]

    for i in range(out_label_ids.shape[0]):
        for j in range(out_label_ids.shape[1]):
            if out_label_ids[i, j] != pad_token_label_id:
                out_label_list[i].append(label_map[out_label_ids[i][j]])
                preds_list[i].append(label_map[preds[i][j]])
                confidence_lst[i].append((logits_store[i - 1][j]))

    results = {
        "loss": eval_loss,
        "precision": precision_score(out_label_list, preds_list),
        "recall": recall_score(out_label_list, preds_list),
        "f1": f1_score(out_label_list, preds_list),
    }
    # print('results==', results)
    # print('prediction_list===',len(preds_list[0]))
    # print('out_label_list===', out_label_list)
    try:
        report = classification_report(out_label_list, preds_list)
        logger.info("\n" + report)
    except Exception as e:
        logger.info("\n" + "no classification report")


    logger.info("***** Eval results %s *****", prefix)
    for key in sorted(results.keys()):
        logger.info("  %s = %s", key, str(results[key]))

    return results, preds_list, confidence_lst

**Comment by kbrajwani** [comment]
Hey @sreejith3534  Thanks for great script.
Can you tell me whey you have taken i - 1 here instead of i .
confidence_lst[i].append((logits_store[i - 1][j]))
